src/communication/communicator.hpp

```diff
@@ -45,31 +48,71 @@ class communicator {
 
     communicator() {}
 
-    explicit communicator(gid_partition_type cell_gid_partition):
-        cell_gid_partition_(cell_gid_partition)
-    {}
-
-    cell_local_size_type num_groups_local() const
-    {
-        return cell_gid_partition_.size();
-    }
-
-    void add_connection(connection con) {
-        EXPECTS(is_local_cell(con.destination().gid));
-        connections_.push_back(con);
-    }
-
-    /// returns true if the cell with gid is on the domain of the caller
-    bool is_local_cell(cell_gid_type gid) const {
-        return algorithms::in_interval(gid, cell_gid_partition_.bounds());
-    }
+    explicit communicator(const recipe& rec, const domain_decomposition& dom_dec) {
+        using util::make_span;
+        num_domains_ = comms_.size();
+        num_local_groups_ = dom_dec.num_local_groups();
+
+        // For caching information about each cell
+        struct gid_info {
+            using connection_list = decltype(rec.connections_on(0));
+            cell_gid_type gid;
+            cell_gid_type local_group;
+            connection_list conns;
+            gid_info(cell_gid_type g, cell_gid_type lg, connection_list c):
+                gid(g), local_group(lg), conns(std::move(c)) {}
+        };
+
+        // Make a list of local gid with their group index and connections
+        //   -> gid_infos
+        // Count the number of local connections (i.e. connections terminating on this domain)
+        //  -> n_cons: scalar
+        // Calculate and store domain id of the presynaptic cell on each local connection
+        //  -> src_domains: array with one entry for every local connection
+        // Also the count of presynaptic sources from each domain
+        //  -> src_counts: array with one entry for each domain
+        std::vector<gid_info> gid_infos;
+        gid_infos.reserve(dom_dec.num_local_cells());
+
+        cell_local_size_type n_cons = 0;
+        std::vector<unsigned> src_domains;
+        std::vector<cell_size_type> src_counts(num_domains_);
+        for (auto i: make_span(0, num_local_groups_)) {
+            const auto& group = dom_dec.get_group(i);
+            for (auto gid: group.gids) {
+                gid_info info(gid, i, rec.connections_on(gid));
+                n_cons += info.conns.size();
+                for (auto con: info.conns) {
+                    const auto src = dom_dec.gid_domain(con.source.gid);
```

Comment:
We know below it has to be promotable to `size_t` (it's used as an index into `src_counts`); it may well be clearer just to use `std::size_t` for `src` here and be explicit.